datalad/utils.py

```diff
@@ -117,6 +117,165 @@ def rmtree(path, chmod_files='auto', *args, **kwargs):
     rotree(path, ro=False, chmod_files=chmod_files)
     shutil.rmtree(path, *args, **kwargs)
 
+
+def is_symlink(file_, good=None):
+    """Check if file_ is a symlink
+
+    Parameters
+    ----------
+    good: None or bool, optional
+        Test not only that it is a symlink, but either it is a good symlink
+        (good=True) or broken (good=False) symlink.  So if good=False,
+        it would return True only if it is a symlink and it is broken.
+    """
+    try:
+        link = os.readlink(file_)
+    except OSError:
+        link = None
+
+    # TODO: copied from ok_symlink so may be we also want to check the path etc?
+    # path_ = realpath(file_)
+    # ok_(path_ != link)
+    # TODO anything else?
+
+    if good is None:
+        # nothing to do more:
+        return bool(link)
+    else:
+        return is_good_symlink(file_)
+
+
+def is_good_symlink(file_):
+    """Return if a file_ is a good symlink (assuming it is a symlink)
+    """
+    return exists(realpath(file_))
+
+def ls_tree(path, files_only=False, relative=True):
+    """Recurse into the directory and return all found files (and dirs if dirs)
+    """
+    out = []
+    for root, dirs, files in os.walk(path):
+        for f in files:
+            out.append(opj(root, f))
+        if not files_only:
+            for d in dirs:
+                out.append(opj(root, d))
+    out = sorted(out)
+    if relative:
+        beg = len(path) + len(os.pathsep)
+        return [o[beg:] for o in out]
+    else:
+        return out
+
+def has_content(file_):
+    """Returns True if a file is not a broken symlink, and if it has content
+    """
+    if not on_windows:
+        if not is_symlink(file_, good=True):
+            return False
+    # verify it has content
+    return os.stat(realpath(file_)).st_size > 0
+
+
+def rm_empties(p, empty_files=None, empty_dirs=None):
+    """Callback for traverse_for_content to be used for do_any, do_none
+
+    To prune empty content
+    """
+    for d in empty_dirs:
+        rmtree(opj(p, d))
+    for f in empty_files:
+        os.unlink(opj(p, f))
+
+
+def traverse_for_content(path,
+                         do_none=None,
+                         do_any=None,
+                         do_all=None,
```

Comment:
Note: I think I will introduce "do_some" which will take current functionality of ```do_any```. And do_any would take precedence over `do_some` and `do_all` (thus either define do_any or any/both of do_some/do_all, not to_any + any of those too).   this would eliminate need to provide the same callback for do_some and do_all  as I had to do now in my test.
Comments are welcome @bpoldrack